implementation/src/main/java/io/smallrye/jwt/auth/principal/DefaultJWTTokenParser.java

```diff
@@ -51,30 +54,84 @@
     private static final Pattern CLAIM_PATH_PATTERN = Pattern.compile("\\/(?=(?:(?:[^\"]*\"){2})*[^\"]*$)");
 
     private volatile VerificationKeyResolver keyResolver;
+    private volatile DecryptionKeyResolver decryptionKeyResolver;
 
     public JwtContext parse(final String token, final JWTAuthContextInfo authContextInfo) throws ParseException {
 
-        try {
-            JwtConsumerBuilder builder = new JwtConsumerBuilder()
-                    .setRequireExpirationTime();
+        String tokenSequence = token;
+        ProtectionLevel level = getProtectionLevel(authContextInfo);
 
-            if (authContextInfo.getMaxTimeToLiveSecs() != null) {
-                builder.setRequireIssuedAt();
-            }
+        if (level == ProtectionLevel.SIGN_ENCRYPT) {
+            tokenSequence = decryptSignedToken(tokenSequence, authContextInfo);
+            level = ProtectionLevel.SIGN;
+        }
+        return parseClaims(tokenSequence, authContextInfo, level);
```

Comment:
Does the case where `ProtectionLevel.ENCRYPT` is the `level` need to be covered?

Reply:
@MikeEdgar Hi, it is covered in the original check [here](https://github.com/smallrye/smallrye-jwt/pull/265/files/e10d56040ae3403dffdbd3aba1eef05c32b49ced#diff-5f4020e440d6d2dd7b7ff27fbe83dd25), so yes, if there is only decryption key configured somewhere then it is a direct decryption (inside `parseClaims`), but if both verification and decryption keys are available then the highlighted code decrypts it first and then the content is nested signed JWT.

All of it can be done dynamically by checking the token structure but the approach is to be conservative - the production deployments have specific security level policies as opposed to "we can process either signed or signed and encrypted or  just encrypted" tokens :-)